### PR TITLE
Adding left and right joins.

### PR DESCRIPTION
**Pull request by tcoupland**
Added specific left and right join fns, and made a default join that is
right join (to match incanter). There is also an optional way to for
callers to supply what they want empty cell values to be.

### CODE REVIEW

File: test/witan/datasets_test.clj [hunk 1]
```diff
@@ -80,20 +87,63 @@
...
+    (testing "joins"
+      (is
+       (= (d [:a :b :c] [[1 2 5] [2 4 1] [3 2 3]])
+          (wds/join ds1 ds2 [:a])))
+      (is
+       (= (d [:a :b :e] [[1 2 8] [2 4 0] [3 2 nil]])
+          (wds/join ds1 ds3 [:a])))
+      (is
+       (= (d [:a :x :y :b] [[1 8 9 2]])
+          (wds/join ds4 ds1 [:a])))))
```
Comment (otfrom) [inline test/witan/datasets_test.clj:131]: is it worth testing with multiple keys in the join?

File: test/witan/datasets_test.clj [hunk 2]
```diff
@@ -80,20 +87,63 @@
...
+      (is
+       (= (ds/dataset [:a :b :a] [[1 2 2] [2 4 nil] [3 2 3]])
+          (wds/left-join ds2 ds1 [[:a] [:c]]))))
+    (testing "right joins"
+      (is
+       (= (d [:a :b :c] [[1 2 5] [2 4 1] [3 2 3]])
+          (wds/right-join ds1 ds2 [:a])))
+      (is
+       (= (d [:a :b :e] [[1 2 8] [2 4 0] [3 2 nil]])
+          (wds/right-join ds1 ds3 [:a])))
```
Comment (otfrom) [inline test/witan/datasets_test.clj]: Have I misunderstood right joins? I would have expected to see tuples keyed by ds3 1, 2, 4

Comment (tcoupland) [inline test/witan/datasets_test.clj]: It's a right fiddle, and i had it wrong before, hoping i've got it right now. Left join means you join onto the left of the right hand dataset. Right join is join onto the right side of the left hand dataset.

File: test/witan/datasets_test.clj (hunk 1 above)
Comment (tcoupland) [inline test/witan/datasets_test.clj:131]: the larger test does do this, joins by [:b :c].

File: test/witan/datasets_test.clj (hunk 2 above)
Comment (tcoupland) [inline test/witan/datasets_test.clj]: gar, now i'm thinking i've got it wrong!

Comment (otfrom) [inline test/witan/datasets_test.clj]: I'd expect the result of this to be: 

`[[:a :b :e] [ [1 2 8] [2 4 0] [4 nil 1] ]`

**Comment by otfrom** [comment]
:shipit:

**Pull request merged**